src/tablespace.c

```diff
@@ -100,37 +100,217 @@ tablespace_tuple_found(TupleInfo *ti, void *data)
 }
 
-Tablespaces *
-tablespace_scan(int32 hypertable_id)
+#define NO_INDEX -1
+
+static int
+tablespace_scan_internal(int indexid,
+						 ScanKeyData *scankey,
+						 int nkeys,
+						 tuple_found_func tuple_found,
+						 tuple_found_func tuple_filter,
+						 void *data,
+						 int limit,
+						 LOCKMODE lockmode)
 {
 	Catalog    *catalog = catalog_get();
-	Tablespaces *tspcs = tablespaces_alloc(TABLESPACE_DEFAULT_CAPACITY);
-	ScanKeyData scankey[1];
 	ScannerCtx	scanctx = {
 		.table = catalog->tables[TABLESPACE].id,
-		.index = catalog->tables[TABLESPACE].index_ids[TABLESPACE_HYPERTABLE_ID_TABLESPACE_NAME_IDX],
-		.scantype = ScannerTypeIndex,
-		.nkeys = 1,
+		.index = (indexid == NO_INDEX) ? 0 : catalog->tables[TABLESPACE].index_ids[indexid],
+		.scantype = (indexid == NO_INDEX) ? ScannerTypeHeap : ScannerTypeIndex,
+		.nkeys = nkeys,
 		.scankey = scankey,
-		.data = tspcs,
-		.tuple_found = tablespace_tuple_found,
-		.lockmode = AccessShareLock,
+		.tuple_found = tuple_found,
+		.filter = tuple_filter,
+		.data = data,
+		.limit = limit,
+		.lockmode = lockmode,
 		.scandirection = ForwardScanDirection,
 	};
 
+	return scanner_scan(&scanctx);
+}
+
+Tablespaces *
+tablespace_scan(int32 hypertable_id)
+{
+	Tablespaces *tspcs = tablespaces_alloc(TABLESPACE_DEFAULT_CAPACITY);
+	ScanKeyData scankey[1];
+
 	ScanKeyInit(&scankey[0], Anum_tablespace_hypertable_id_tablespace_name_idx_hypertable_id,
 				BTEqualStrategyNumber, F_INT4EQ, Int32GetDatum(hypertable_id));
 
-	scanner_scan(&scanctx);
+	tablespace_scan_internal(TABLESPACE_HYPERTABLE_ID_TABLESPACE_NAME_IDX,
+							 scankey,
+							 1,
+							 tablespace_tuple_found,
+							 NULL,
+							 tspcs,
+							 0,
+							 AccessShareLock);
 
 	return tspcs;
 }
 
+
+typedef struct TablespaceScanInfo
+{
+	Catalog    *catalog;
+	Cache	   *hcache;
+	Oid			userid;
+	int			num_filtered;
+	int			stopcount;
+	void	   *data;
+} TablespaceScanInfo;
+
+static int
+tablespace_scan_by_name(const char *tspcname, tuple_found_func tuple_found, void *data)
+{
+	ScanKeyData scankey[1];
+	int			nkeys = 0;
+
+	if (NULL != tspcname)
+		ScanKeyInit(&scankey[nkeys++],
+					Anum_tablespace_tablespace_name,
+					BTEqualStrategyNumber, F_NAMEEQ,
+					DirectFunctionCall1(namein, CStringGetDatum(tspcname)));
+
+	return tablespace_scan_internal(NO_INDEX,
+									scankey,
+									nkeys,
+									tuple_found,
+									NULL,
+									data,
+									0,
+									AccessShareLock);
+}
+
+int
+tablespace_is_attached(const char *tspcname)
+{
+	return tablespace_scan_by_name(tspcname, tablespace_tuple_found, NULL);
+}
+
+static void
+tablespace_validate_revoke_internal(const char *tspcname, tuple_found_func tuple_found, void *stmt)
+{
+	TablespaceScanInfo info = {
+		.catalog = catalog_get(),
+		.hcache = hypertable_cache_pin(),
+		.data = stmt,
+	};
+
+	tablespace_scan_by_name(tspcname, tuple_found, &info);
+
+	cache_release(info.hcache);
+}
+
+static void
+validate_revoke(Oid tspcoid, Oid role, Oid relid)
```

Comment:
That would be confusing I think, since it would imply a validation for creating the tablespace, not for revoking the permission `ACL_CREATE`. Maybe `validate_revoke_create()`?